dtm0/recovery.c

```diff
@@ -0,0 +1,2170 @@
+						 M0_CST_DTM0, &svc_fid);
+		if (rc != 0)
+			continue;
+
+		rf = recovery_fom_by_svc_find(m, &svc_fid);
+		if (rf == NULL)
+			break;
+
+		if (!rf->rf_is_local)
+			heq_post(rf, proc_conf->pc_obj.co_ha_state);
+	}
+
+	m0_conf_diter_fini(&it);
+	/* end of workaround */
+
+out:
+	if (root != NULL)
+		m0_confc_close(&root->rt_obj);
+	if (rc != 0)
+		unpopulate_foms(m);
+	return M0_RC(rc);
+}
+
+static struct m0_co_context *CO(struct m0_fom *fom)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+	return &rf->rf_coro;
+}
+
+#define F M0_CO_FRAME_DATA
+
+static void heq_await(struct m0_fom *fom, enum m0_ha_obj_state *out, bool *eoq)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+
+	M0_CO_REENTER(CO(fom),
+		      struct m0_be_op op;
+		      bool            got;
+		      uint64_t        state;);
+
+	F(got) = false;
+	F(state) = 0;
+	M0_SET0(&F(op));
+	m0_be_op_init(&F(op));
+	m0_be_queue_lock(&rf->rf_heq);
+	M0_BE_QUEUE_GET(&rf->rf_heq, &F(op), &F(state), &F(got));
+	m0_be_queue_unlock(&rf->rf_heq);
+	M0_CO_YIELD_RC(CO(fom), m0_be_op_tick_ret(&F(op), fom, RFS_WAITING));
+	m0_be_op_fini(&F(op));
+
+
+	if (F(got)) {
+		M0_LOG(M0_DEBUG, "heq_deq " FID_F " %s" ,
+		       FID_P(&rf->rf_tgt_svc),
+		       m0_ha_state2str(F(state)));
+		M0_ASSERT(ha_event_invariant(F(state)));
+		*out = F(state);
+	} else {
+		M0_LOG(M0_DEBUG, "heq_deq " FID_F, FID_P(&rf->rf_tgt_svc));
+		*eoq = true;
+	}
+}
+
+static void eolq_await(struct m0_fom *fom, struct eolq_item *out)
+{
+	struct recovery_fom *rf = M0_AMB(rf, fom, rf_base);
+
+	M0_CO_REENTER(CO(fom),
+		      struct m0_be_op op;
+		      bool            got;
+		      struct eolq_item item;);
+
+	F(got) = false;
+	M0_SET0(&F(item));
+	M0_SET0(&F(op));
+	m0_be_op_init(&F(op));
+	m0_be_queue_lock(&rf->rf_eolq);
+	M0_BE_QUEUE_GET(&rf->rf_eolq, &F(op), &F(item), &F(got));
+	m0_be_queue_unlock(&rf->rf_eolq);
+	M0_CO_YIELD_RC(CO(fom), m0_be_op_tick_ret(&F(op), fom, RFS_WAITING));
+	m0_be_op_fini(&F(op));
+
+	*out = F(got) ? F(item) : (struct eolq_item) { .ei_type = EIT_END };
+}
+
+static void restore(struct m0_fom *fom,
+		    enum m0_ha_obj_state *out, bool *eoq)
+{
+	struct recovery_fom   *rf = M0_AMB(rf, fom, rf_base);
+	struct m0_dtm0_log_rec record;
+	int                    rc = 0;
+
+	M0_CO_REENTER(CO(fom),
+		      struct dtm0_req_fop redo;
+		      struct m0_fid       initiator;
+		      struct m0_be_op     reply_op;
+		      bool                next;
+		      );
+
+	recovery_machine_log_iter_init(rf->rf_m, &rf->rf_log_iter);
+
+	/* XXX: race condition in the case where we are stopping the FOM. */
+	F(initiator) = recovery_fom_local(rf->rf_m)->rf_tgt_svc;
+
+	M0_SET0(&F(reply_op));
+	m0_be_op_init(&F(reply_op));
+
+	do {
+		M0_SET0(&record);
+		rc = recovery_machine_log_iter_next(rf->rf_m, &rf->rf_log_iter,
+						    &rf->rf_tgt_svc, NULL,
+						    &record);
+		/* Any value except zero means that we should stop recovery. */
+		F(next) = rc == 0;
+		F(redo) = (struct dtm0_req_fop) {
```

Comment:
I am new to the coroutine programming, but I have a question here.
The 'redo' is initialized to new content in every iteration, and 'redo' is not used after "M0_CO_YIELD()".
So, it can be declared as a normal local variable, like `record`?

Reply:
3:Working on it.

You're right, the way it's written now, it can be local.  This was originally put into coroutine context because it was not clear when and how payload duplicates will be made -- so it was kept in context, to make sure it's not destroyed before it's actually sent.

Reply:
Fixed/merged.